media/js/firefox/hello/hello-ftu.js

```diff
@@ -0,0 +1,471 @@
+
+// create namespace
+if (typeof Mozilla === 'undefined') {
+    var Mozilla = {};
+}
+
+(function($) {
+    'use strict';
+
+    var HelloFTU = {};
+    var _highlightTimeout;
+    var $defaultCopy = $('.default');
+    var $startCopy = $('.start');
+    var $endCopy = $('.end');
+
+    HelloFTU.highlightSupressed = false;
+
+    // Used for unit testing purposes only as "document.hidden" is read only.
+    HelloFTU.documentHidden = null;
+
+    HelloFTU.tourSource = 'none';
+
+    /**
+     * Initialize Hello FTU.
+     * Only start the tour if the Hello icon is present in the tool pallet.
+     * Given that this tour should only be accessed via the Hello panel itself,
+     * this should always be "true". However, it's always possible someone may
+     * share the URL directly with someone who has disabled or removed Hello.
+     */
+    HelloFTU.init = function() {
+        // Hello is still referred to as 'loop' internally for legacy reasons.
+        Mozilla.UITour.getConfiguration('availableTargets', function (config) {
+            if (config.targets && $.inArray('loop', config.targets) !== -1) {
+                HelloFTU.tourStep = HelloFTU.getInitialTourStep();
+                HelloFTU.setInitialPageState();
+                HelloFTU.bindPageEvents();
+                HelloFTU.bindHelloEvents();
+                HelloFTU.trackTourSource();
+                HelloFTU.showTourStep();
+            }
+        });
+    };
+
+    HelloFTU.getInitialTourStep = function() {
+        // URL query param populated at template level in the view
+        var incomingConversation = HelloFTU.getParameterByName('incomingConversation');
+        var tourStep = 'get-started';
+
+        // set the tour state based on incomingConversation status
+        switch(incomingConversation) {
+        case 'waiting':
+            tourStep = 'conversation-waiting';
+            break;
+        case 'open':
+            tourStep = 'conversation-open';
+            break;
+        }
+
+        return tourStep;
+    };
+
+    HelloFTU.setInitialPageState = function() {
+        if (HelloFTU.tourStep === 'conversation-open') {
+            HelloFTU.showPageState('end');
+            HelloFTU.replaceURLState('open', 'done');
+            HelloFTU.trackGAConversationConnect();
+        } else {
+            HelloFTU.showPageState('start');
+        }
+    };
+
+    /*
+     * Show the Hello panel and trigger callback once opening animation finishes
+     * @param callback (function)
+     */
+    HelloFTU.showHelloPanel = function(callback) {
+        // Only open the info panel if tab is visible
+        if (HelloFTU.documentHidden || document.hidden) {
+            return;
+        }
+        // Make sure loop icon is available before opening Hello panel (bug 1111828)
+        Mozilla.UITour.getConfiguration('availableTargets', function (config) {
+            if (config.targets && $.inArray('loop', config.targets) !== -1) {
```

Comment:
Is this operation expensive? Is it feasible to cache the result of this check so it only has to be executed once? I suppose the user *could* move the icon between page load and initializing the tour...